### PR TITLE
Add `constants/` crate

### PR DESCRIPTION
**Pull request by hinto-janai**
### What
Adds a `cuprate_constants` crate in the root `constants/` folder.

Closes https://github.com/Cuprate/cuprate/issues/189.

### Why, where
See #189.

### How
See below review.

### CODE REVIEW

**Comment by Asurar0** [comment]
Will P2P constants be included in this crate as well?

https://github.com/Cuprate/cuprate/blob/main/p2p/p2p-core/src/constants.rs
https://github.com/Cuprate/cuprate/blob/main/p2p/p2p/src/constants.rs

**Comment by hinto-janai** [comment]
> Will P2P constants be included in this crate as well?

I haven't checked in depth but if they are `pub(crate)` and not needed elsewhere, probably not. Constants like `CRYPTONOTE_MAX_BLOCK_HEIGHT` doesn't really belong anywhere yet is needed by multiple crates - these are the type of constants this crate will contain.

File: pruning/src/lib.rs [hunk 1]
```diff
@@ -20,9 +20,10 @@
 
 use std::cmp::Ordering;
 
+use cuprate_constants::block::MAX_BLOCK_HEIGHT_USIZE;
+
 use thiserror::Error;
 
-pub const CRYPTONOTE_MAX_BLOCK_HEIGHT: usize = 500000000;
```
Comment (hinto-janai) [inline pruning/src/lib.rs:25]: `CRYPTONOTE_MAX_BLOCK_HEIGHT` is replaced by `cuprate_constants::block::MAX_BLOCK_HEIGHT[_USIZE]` throughout the codebase.

Also, I think Cuprate is far enough removed from the original CryptoNote that we can start dropping the `CRYPTONOTE_` prefix.

File: constants/src/genesis.rs [hunk 2]
```diff
@@ -0,0 +1,177 @@
...
+    }
+
+    Stagenet {
+        nonce: 10002,
+        block: "010000000000000000000000000000000000000000000000000000000000000000000012270000013c01ff0001ffffffffffff0302df5d56da0c7d643ddd1ce61901c7bdc5fb1738bfe39fbe69c28a3a7032729c0f2101168d0c4ca86fb55a4cf6a36d31431be1c53a3bd7411bb24e8832410289fa6f3b00",
+        block_hash: "76ee3cc98646292206cd3e86f74d88b4dcc1d937088645e9b0cbca84b7ce74eb",
+        tx: "013c01ff0001ffffffffffff0302df5d56da0c7d643ddd1ce61901c7bdc5fb1738bfe39fbe69c28a3a7032729c0f2101168d0c4ca86fb55a4cf6a36d31431be1c53a3bd7411bb24e8832410289fa6f3b",
+        tx_hash: "c099809301da6ad2fde11969b0e9cb291fc698f8dc678cef00506e7baf561de4",
+    }
+}
```
[the remark this answers is not in the document]
Comment (Boog900) [reply constants/src/genesis.rs:177]: As much as this is cool, do we actually need this much info?

The consensus-rules crate has a genesis module which is pretty much this just less complex

[the remark this answers is not in the document]
Comment (hinto-janai) [reply constants/src/genesis.rs:177]: The DB needs access to genesis data and [RPC could benefit from access to the genesis hash](https://github.com/monero-project/monero/blob/master/src/cryptonote_core/blockchain.cpp#L2434).

I know `cuprate_consensus_rules::genesis` exists but it only contains a single non-const function that generates a `Block` per call.

[the remark this answers is not in the document]
Comment (hinto-janai) [reply constants/src/genesis.rs]: > This request is already impl'd in the DB for the block downloader

We could check the hash at the RPC level before touching the DB, anyway, removed the module. If I find more usecases I'd like to add this back.

File: constants/src/block.rs [hunk 3]
```diff
@@ -0,0 +1,25 @@
...
+#[doc = monero_definition_link!(a1dc85c5373a30f14aaf7dcfdd95f5a7375d3623, "/src/cryptonote_config.h", 81)]
+///
+/// ref: <https://monero-book.cuprate.org/consensus_rules/blocks/difficulty.html#target-seconds>
+pub const BLOCK_TIME_V1: Duration = Duration::from_secs(60);
+
+/// Target block time from hardfork v2.
+#[doc = monero_definition_link!(a1dc85c5373a30f14aaf7dcfdd95f5a7375d3623, "/src/cryptonote_config.h", 80)]
+///
+/// ref: <https://monero-book.cuprate.org/consensus_rules/blocks/difficulty.html#target-seconds>
+pub const BLOCK_TIME_V2: Duration = Duration::from_secs(120);
```
Comment (hinto-janai) [inline constants/src/block.rs:25]: Used in `types/src/hard_fork.rs`.

I think this also needs to be used here? https://github.com/Cuprate/cuprate/pull/300/commits/7902f0a902f1caece3a1a12e268ef97fe13df59a#diff-bdf64dc01a71e5057249b389c20030fd2f8db7be263cee26318bd976500f075dL112-R114

File: constants/src/output.rs [hunk 4]
```diff
@@ -0,0 +1,11 @@
...
+
+use crate::macros::monero_definition_link;
+
+/// The minimum amount of blocks a coinbase output is locked for.
+#[doc = monero_definition_link!(a1dc85c5373a30f14aaf7dcfdd95f5a7375d3623, "/src/cryptonote_config.h", 40)]
+pub const COINBASE_LOCK_WINDOW: usize = 60;
+
+/// The minimum amount of blocks an output is locked for.
+#[doc = monero_definition_link!(a1dc85c5373a30f14aaf7dcfdd95f5a7375d3623, "/src/cryptonote_config.h", 49)]
+pub const DEFAULT_LOCK_WINDOW: usize = 10;
```
Comment (hinto-janai) [inline constants/src/output.rs:11]: Also used in https://github.com/Cuprate/cuprate/pull/300/commits/7902f0a902f1caece3a1a12e268ef97fe13df59a#diff-bdf64dc01a71e5057249b389c20030fd2f8db7be263cee26318bd976500f075dL112-R114.

**Comment by Boog900** [comment]
I don't think we should include block time/output lock time constants here, unless you need to use them elsewhere?

**Comment by hinto-janai** [comment]
@Boog900 ready.

**Comment by hinto-janai** [comment]
Actually, we need block time targets here: https://github.com/monero-project/monero/blob/9866a0e9021e2422d8055731a586083eb5e2be67/src/rpc/core_rpc_server.cpp#L531

**Review by Boog900 — APPROVED** [review]

**Pull request merged**